iOS builds fail when using `koinViewModel` in compose multiplatform: `Undefined symbols for architecture arm64`

**Describe the bug**
When using `val viewModel = koinViewModel<VM>()` in a compose multiplatform project the build for iOS fails, with the following error message:

```
Undefined symbols for architecture arm64:
  "_kfun:androidx.lifecycle.viewmodel.compose#androidx_lifecycle_viewmodel_compose_LocalViewModelStoreOwner$stableprop_getter$artificial(){}kotlin.Int"
```

After commenting out all usages of `val viewModel = koinViewModel<VM>()` the app builds fine for iOS, so the issue is with `koinViewModel`.

Android builds always work, with or without `koinViewModel`.

**Context:**
Using the jetbrains navigation and jetbrains viewmodel that was ported from android for compose multiplatform, e.g. like so for example:

```
val navController = rememberNavController()
NavHost(
    navController = navController,
    startDestination = DashboardRoute,
) {
    composable<DashboardRoute> {
        val viewModel = koinViewModel<DashboardViewModel>() // This prevents iOS builds
        DashboardScreen(
            viewModel = viewModel,
            navController = navController,
        )
    }
}
```

Using `val viewModel = viewModel { DashboardViewModel() }`, e.g. the default Google way, instead works fine.

**Koin module and version:**
```
[versions]
koin = "4.0.4"

[libraries]
koin-core = { module = "io.insert-koin:koin-core", version.ref = "koin" }
koin-compose = { module = "io.insert-koin:koin-compose", version.ref = "koin" }
koin-compose-viewmodel = { module = "io.insert-koin:koin-compose-viewmodel", version.ref = "koin" }
koin-compose-navigation = { module = "io.insert-koin:koin-compose-viewmodel-navigation", version.ref = "koin" }
```

I had the exact same issue when build on iOS
> Undefined symbols for architecture arm64:  "_kfun:androidx.lifecycle.viewmodel.compose#androidx_lifecycle_viewmodel_compose_LocalViewModelStoreOwner$stableprop_getter$artificial(){}kotlin.Int"

I found that adding the following line to my gradle.properties completely solved the issue.

**gradle.properties:**
```
kotlin.native.cacheKind=none
```

This is likely related to [JetBrains issue CMP-7571](https://youtrack.jetbrains.com/issue/CMP-7571), which explains incompatibilities between K1/K2 compiled libraries in Compose Multiplatform and Kotlin 2.1.x.

I hope this helps you!

> I had the exact same issue when build on iOS
> 
> > Undefined symbols for architecture arm64:  "_kfun:androidx.lifecycle.viewmodel.compose#androidx_lifecycle_viewmodel_compose_LocalViewModelStoreOwner$stableprop_getter$artificial(){}kotlin.Int"
> 
> I found that adding the following line to my gradle.properties completely solved the issue.
> 
> **gradle.properties:**
> 
> ```
> kotlin.native.cacheKind=none
> ```
> 
> This is likely related to [JetBrains issue CMP-7571](https://youtrack.jetbrains.com/issue/CMP-7571), which explains incompatibilities between K1/K2 compiled libraries.
> 
> I hope this helps you!

Thanks for the tip @bvic-dev, iOS builds work for me when adding the fix you have provided!

--- a/projects/compose/koin-compose-viewmodel-navigation/build.gradle.kts
+++ b/projects/compose/koin-compose-viewmodel-navigation/build.gradle.kts
@@ -1,8 +1,5 @@
-import org.jetbrains.kotlin.gradle.dsl.JvmTarget
-import org.jetbrains.kotlin.gradle.targets.js.nodejs.NodeJsRootExtension
-import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
-
 plugins {
+    alias(libs.plugins.androidLibrary)
     alias(libs.plugins.kotlinMultiplatform)
     alias(libs.plugins.jetbrainsCompose)
     alias(libs.plugins.compose.compiler)
@@ -12,9 +9,17 @@ val koinVersion: String by project
 version = koinVersion
 
 kotlin {
-    
+    androidTarget {
+        publishLibraryVariants("release")
+        compilations.all {
+            kotlinOptions.jvmTarget = "1.8"
+        }
+    }
+
     jvm {
-        withJava()
+        compilations.all {
+            kotlinOptions.jvmTarget = "11"
+        }
     }
 
     js(IR) {
@@ -37,16 +42,16 @@ kotlin {
     sourceSets {
         commonMain.dependencies {
             api(project(":compose:koin-compose-viewmodel"))
-//            api(project(":core:koin-core-viewmodel-navigation"))
             api(libs.jb.composeNavigation)
         }
     }
 }
 
-tasks.withType<KotlinCompile>().all {
-    compilerOptions {
-        jvmTarget.set(JvmTarget.JVM_1_8)
-    }
+val androidCompileSDK: String by project
+
+android {
+    namespace = "org.koin.compose.viewmodel.navigation"
+    compileSdk = androidCompileSDK.toInt()
 }
 
 tasks.withType<org.jetbrains.kotlin.gradle.targets.js.npm.tasks.KotlinNpmInstallTask>().configureEach {

--- a/projects/compose/koin-compose-viewmodel/build.gradle.kts
+++ b/projects/compose/koin-compose-viewmodel/build.gradle.kts
@@ -1,8 +1,5 @@
-import org.jetbrains.kotlin.gradle.dsl.JvmTarget
-import org.jetbrains.kotlin.gradle.targets.js.nodejs.NodeJsRootExtension
-import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
-
 plugins {
+    alias(libs.plugins.androidLibrary)
     alias(libs.plugins.kotlinMultiplatform)
     alias(libs.plugins.jetbrainsCompose)
     alias(libs.plugins.compose.compiler)
@@ -12,9 +9,17 @@ val koinVersion: String by project
 version = koinVersion
 
 kotlin {
-    
+    androidTarget {
+        publishLibraryVariants("release")
+        compilations.all {
+            kotlinOptions.jvmTarget = "1.8"
+        }
+    }
+
     jvm {
-        withJava()
+        compilations.all {
+            kotlinOptions.jvmTarget = "11"
+        }
     }
 
     js(IR) {
@@ -43,10 +48,11 @@ kotlin {
     }
 }
 
-tasks.withType<KotlinCompile>().all {
-    compilerOptions {
-        jvmTarget.set(JvmTarget.JVM_1_8)
-    }
+val androidCompileSDK: String by project
+
+android {
+    namespace = "org.koin.compose.viewmodel"
+    compileSdk = androidCompileSDK.toInt()
 }
 
 tasks.withType<org.jetbrains.kotlin.gradle.targets.js.npm.tasks.KotlinNpmInstallTask>().configureEach {

--- a/projects/compose/koin-compose/build.gradle.kts
+++ b/projects/compose/koin-compose/build.gradle.kts
@@ -1,9 +1,3 @@
-import org.gradle.kotlin.dsl.android
-import org.gradle.kotlin.dsl.api
-import org.gradle.kotlin.dsl.jvm
-import org.jetbrains.kotlin.gradle.dsl.JvmTarget
-import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
-
 plugins {
     alias(libs.plugins.androidLibrary)
     alias(libs.plugins.kotlinMultiplatform)
@@ -15,10 +9,17 @@ val koinVersion: String by project
 version = koinVersion
 
 kotlin {
-    
-    jvm()
     androidTarget {
         publishLibraryVariants("release")
+        compilations.all {
+            kotlinOptions.jvmTarget = "1.8"
+        }
+    }
+
+    jvm {
+        compilations.all {
+            kotlinOptions.jvmTarget = "11"
+        }
     }
 
     js(IR) {
@@ -73,12 +74,6 @@ android {
     compileOptions {
         sourceCompatibility = JavaVersion.VERSION_1_8
         targetCompatibility = JavaVersion.VERSION_1_8
-    }
-}
-
-tasks.withType<KotlinCompile>().all {
-    compilerOptions {
-        jvmTarget.set(JvmTarget.JVM_1_8)
     }
 }
 

--- a/projects/gradle/libs.versions.toml
+++ b/projects/gradle/libs.versions.toml
@@ -22,9 +22,9 @@ androidx-startup= "1.2.0"
 
 # Compose
 composeJetpackRuntime = "1.7.8"
-composeJB = "1.7.3"
+composeJB = "1.8.0-rc01"
 composeLifecycle = "2.8.4"
-composeNavigation = "2.8.0-alpha10" # 2.8.0-alpha02 to keep compose 1.6.11
+composeNavigation = "2.9.0-alpha17"
 jbCoreBundle = "1.0.1"
 jbSavedState = "1.2.2"
 

